src/crimson/os/poseidonstore/poseidonstore.h

```diff
@@ -0,0 +1,234 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <map>
+#include <typeinfo>
+#include <vector>
+
+#include <optional>
+#include <seastar/core/future.hh>
+
+#include "osd/osd_types.h"
+#include "include/uuid.h"
+
+#include "os/Transaction.h"
+#include "crimson/os/futurized_store.h"
+#include "transaction.h"
+#include "wal.h"
+#include "cache.h"
+
+namespace crimson::os::poseidonstore {
+
+/**
+ *
+ * PoseidonStore write processing
+ *
+ * 1) Thread architecture and lock domain
+ *
+ * Poseidonstore aims to split a partition into several partitions to maximize parallelism based on Crimson, 
+ * but the number of partitionss is smaller than the number of PGs in an OSD (PGs >= Partitions). 
+ * So, the reactor threads, which handle I/Os based on per-PG basis, inevitabely access to the partition 
+ * at the same time. The following figure  shows the example.
+ *
+ *
+ * (PG 1) reactor thread  (PG 2) reactor thread   (PG 3) reactor thread
+ *      |                  |----------------- |
+ *      ----------------------|
+ *           V-------------|  V
+ * ------------------------------------
+ * | partition 1    || partition 2    |
+ * | (PG 2, 3)      || (PG 1)         |
+ * ====================================
+ * | WAL            || WAL            |
+ * ====================================
+ * | Data partition || Data partition |
+ * ------------------------------------
+ * partion 1 means a sharded partition
+ * 
+ * ** Discussion
+ * The best way to achieve high IOPS is that each sharded service (reactor thread) manages only 
+ * a sharded partition. With this policy; each sharded partion handles pre-configured PGs,
+ * if the number of PGs grows, the number of sharded parition can also increase. However, what happens if
+ * the number of sharded services (reactor thread) grows or shrinks with the same PGs? 
+ * Do we need to consider this case?
+ * To be conservative, the following design assumes that we need a lock to protect two reactor thread
+ * access a shared partition at a time, considering such case.
```

Comment:
This discussion really isn't useful here.  Please replace it with a concise, exact description of what you are doing.